fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -946,6 +961,120 @@ public void rebuildIndex(@Nonnull FDBRecordStore store) {
         return buildAsync(this::buildEndpoints, false, additionalLogMessageKeyValues);
     }
 
+    /**
+     * This defines in which situation the index should be built. {@link #BUILD_IF_DISABLED},
+     * {@link #BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY}, {@link #BUILD_IF_DISABLED_REBUILD_IF_WRITE_ONLY}, and
+     * {@link #BUILD_IF_DISABLED_REBUILD_IF_WRITE_ONLY_OR_READABLE} are sorted in a way that each one is more exhaustive
+     * than the ones before it.
+     * <p>
+     * In which, {@link #BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY} is recommended if there is no reason to believe
+     * current index data is corrupted.
+     * </p>
+     */
+    public enum BuildOption {
+        /**
+         * Only build if the index is disabled.
+         */
+        BUILD_IF_DISABLED,
+        /**
+         * Build if the index is disabled; Continue build if the index is write-only.
+         * <p>
+         * Recommended. This should be sufficient if current index data is not corrupted.
+         * </p>
+         */
+        BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY,
+        /**
+         * Build if the index is disabled; Rebuild if the index is write-only.
+         */
+        BUILD_IF_DISABLED_REBUILD_IF_WRITE_ONLY,
+        /**
+         * Rebuild the index anyway.
+         */
+        BUILD_IF_DISABLED_REBUILD_IF_WRITE_ONLY_OR_READABLE,
+        /**
+         * Build if the index is disabled; Continue build if the index is write-only, but rebuild if the index is
+         * readable.
+         * <p>
+         * This option should not be considered. It does not make much sense in any use case can be thought of. It is
+         * provided just for completeness.
+         * </p>
+         */
+        BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY_REBUILD_IF_READABLE,
+    }
+
+
+    /**
+     * Builds an index across multiple transactions safely. It differs from {@link #buildIndexAsync()} in two ways:
+     * <ul>
+     *     <li>
+     *         It stops with {@link com.apple.foundationdb.synchronizedsession.SynchronizedSessionLockedException} if
+     *         there is another runner actively working on the same index.
+     *     </li>
+     *     <li>
+     *         It checks and updates index states and clear index data respecting the {@link BuildOption}. So there is
+     *         no need to change the index states or clear existing index data before invoking this method.
+     *     </li>
+     * </ul>
+     * @return a future that will be ready when the build has completed.
+     * @throws com.apple.foundationdb.synchronizedsession.SynchronizedSessionLockedException the build is stopped
+     * because there may be another build running actively on this index.
+     */
+    // TODO: It should support range build (for building index in parallel) as well.
+    @Nonnull
+    public CompletableFuture<Void> safelyBuildIndexAsync() {
```

Comment:
Yes, I agree we should consolidate the two APIs to one. As for what should be provides and what should be default, I believe "safely" here are actually two independent features.

1) non-safe version does not use synchronized session, safe version uses it.

2) non-safe version assumes the index is write-only and existing entries are clear, safe version does not assume that but uses `BuildOption` to decide do this.

Though I think the safe version should be used in any situation for both features in the long term, it makes sense to keep the non-safe version because of reasons.

For 1) the major concern is synchronized sessions may have some unexpected problems and we want to have a way to fall back. So the non-safe version should be provided but not by default. Adding an immutable option in `OnlineIndexer` to opt-out should be sufficient. 

For 2) the major concern is the contract is changed - the clients may be expecting an error when the index is disabled, which may be morbid but may need to be respected. It is perfectly reasonable to add a special build option for the old behavior but I'm reluctant to make it default (because `BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY` is the future). But the problem of not making the old behavior as default is that it is a breaking and confusing change. I'm thinking to keep `buildIndexAsync` behave same as before (by using this build option) but mark it as deprecated. And have a new method to do the right things (what should its name be?).

I'm trying to add options to `OnlineIndexer` builder but not `buildIndexAsync` methods, because the latter may introduce more fragments after we have safe range builds.

I haven't made change for this yet. What do you think? @scgray @alecgrieser

Reply:
I think the flag you mentioned for (1) sounds good. I also agree with the general principle of putting options in the builder rather than on the method.

As to (2), I'm not sure that making it start working on `DISABLED` indexes instead of throwing an error is necessarily a bad change (like, people relying on that might have been doing things wrong). Another option would be to make the default `IndexStatePrecondition` a new option that is something like `ERROR_IF_DISABLED_CONTINUE_IF_WRITE_ONLY` which I think is exactly the current behavior. And then we could document that that is the default for compatibility reasons, but most users should make `BUILD_IF_DISABLED_CONTINUE_IF_WRITE_ONLY` their default option. We could even update the default to the latter (with proper documentation) at some point (I think).

I *think* that would be my preference over a new method, but I could be wrong. I suppose we should also clarify in the docs what are plan is for the various range-based methods if we decided to go with this. (And/or, do the range-based options need to be updated in this PR in that case?)

Reply:
My $0.02...

I do think the option to disable the "safe" build mode is still important and should be added (but have it default to always be safe).   

I think `BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY` is the right thing to do.  It *technically* is a breaking change, but I am not convinced that it could break anything in a truly negative fashion.  As @alecgrieser suggested, we could include `ERROR_IF_DISABLED_CONTINUE_IF_WRITE_ONLY` as a way of restoring the default behavior for the truly wacko client that thought it was a good thing and relied on it (sorry, if you are the wacko I am referring to and reading this commentary).